feat: Add support for UUID Version 8 (RFC 9562)

ashwingopalsamy opened the pull request:
### Changes 

This adds support for generating UUID Version 8 as defined in [RFC 9562](https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-8). UUIDv8 is designed for custom or experimental use cases, allowing you to embed application-specific data while adhering to standard version and variant bits.

**What’s included:**

- `NewV8()`: Generates a UUIDv8 with random values.
- `NewV8FromReader(customA, customB, r)`: Allows custom values for `custom_a` (48 bits) and `custom_b` (12 bits) while `custom_c` field (62 bits) is filled using a random source or secure fallback.
- `NewV8TimeBased(r)`: A time-based variant of `UUIDv8` that uses the current timestamp for `custom_a`. It also includes a sequence number to ensure uniqueness when multiple UUIDs are generated in the same nanosecond.

This addition completes support for all UUID versions in RFC 9562. Looking forward for your feedback, team!

ashwingopalsamy committed:
feature: add UUID version8 support

ashwingopalsamy committed:
tests: add UUID version8 tests

ashwingopalsamy reviewed (commented):
Hi @bormanp, apologies for the noise! I’d really appreciate it if you could take a look at this proposal when you have a moment. Thanks in advance! 🙏

vtolstov commented:
any news ? uuid v8 really useful when need to store in db and have ordered records with uniq node id

ashwingopalsamy commented:
hi team, would appreciate your eyes on this proposal.

@quartzmo /  @bormanp / @noahdietz :)

ashwingopalsamy commented:
Hello 👋 Would appreciate some review here :) 

@bormanp

cristaloleg commented on version8.go:
```diff
@@ -0,0 +1,150 @@
+	timestamp := uint64(time.Now().UnixNano())
+
+	timeMu.Lock()
+	defer timeMu.Unlock()
```
Can we release lock after the `if` below?

ashwingopalsamy replied:
🤔 Sure, I can do that.

cristaloleg replied:
Not like that. Drop `defer`, it's not needed after the `if`.

ashwingopalsamy replied:
Alrights! done.

cristaloleg replied:
`defer`, not `mu.Unlock()`

cristaloleg commented on version8.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+// makeV8 generates a version 8 UUID using user-provided or random data for custom_a, custom_b, and custom_c.
+func makeV8(uuid []byte, customA, customB, customC []byte) {
```
Not used?

ashwingopalsamy replied:
Ah yes! This wasn't supposed to be here. Removed it.

ashwingopalsamy committed:
fix: address code review comments

remove redundant function (dead code) and place the `sync.Mutex` lock after the `if` condition

ashwingopalsamy committed:
fix: remove mu.Unlock

arlanram commented:
any news? this feature is so required today

ashwingopalsamy committed:
Update version8.go

ashwingopalsamy commented:
hey @arlanram, its just that we're awaiting for review from maintainers. 

@quartzmo / @bormanp / @noahdietz / @pborman 👀

vtolstov commented:
any news ?

ashwingopalsamy commented:
Hello, @quartzmo / @bormanp / @noahdietz / @pborman. Sorry for the noice. We'd really appreciate your review here.

vtolstov commented:
@cristaloleg

cristaloleg commented:
@vtolstov I'm just a random guest here

vtolstov commented:
gentle ping

srfrog commented:
what's going, any updates?

jufemaiz commented:
Wondering if Google has just dropped support of this package (and more) as they chase AI sparkles?
